operator/internal/central/extensions/reconcile_tls.go

```diff
@@ -218,6 +275,103 @@ func validateOrGenerateCA(oldCA mtls.CA, oldFileMap types.SecretDataMap) (mtls.C
+			return errors.Wrap(err, "invalid secondary CA in the existing secret, please delete it to allow re-generation")
+		}
+
+		newFileMap[mtls.SecondaryCACertFileName] = caCert
+		newFileMap[mtls.SecondaryCAKeyFileName] = caKey
+		return nil
+	} else if caCertPresent || caKeyPresent {
+		const msg = "malformed secret (%s present but %s missing), please delete it to allow re-generation"
+		if !caCertPresent {
+			return fmt.Errorf(msg, mtls.CAKeyFileName, mtls.CACertFileName)
+		}
+		return fmt.Errorf(msg, mtls.CACertFileName, mtls.CAKeyFileName)
+	}
+
+	return nil
+}
+
+func (r *createCentralTLSExtensionRun) checkCARotation(primary, secondary *x509.Certificate) (CARotationAction, error) {
+	if primary == nil {
+		return CARotateNoAction, errors.New("primary CA must not be nil")
+	}
+
+	startTime := primary.NotBefore
+	endTime := primary.NotAfter
+	if !endTime.After(startTime) {
+		return CARotateNoAction, fmt.Errorf("certificate expires at %s before it begins to be valid at %s", endTime, startTime)
+	}
+	if r.currentTime.Before(startTime) {
+		return CARotateNoAction, fmt.Errorf("certificate lifetime start %s is in the future", startTime)
+	}
+	if r.currentTime.After(endTime) {
+		return CARotateNoAction, fmt.Errorf("certificate expired at %s", endTime)
+	}
+
+	validityDuration := endTime.Sub(startTime)
+	fifthOfValidityDuration := time.Duration(validityDuration.Nanoseconds()/5) * time.Nanosecond
+
+	// Add secondary CA after 3/5 of validity
+	addSecondaryCATime := startTime.Add(3 * fifthOfValidityDuration)
+	if r.currentTime.After(addSecondaryCATime) && secondary == nil {
+		return CARotateAddSecondary, nil
+	}
+
+	// Promote secondary to primary in final year
+	promoteSecondaryCATime := startTime.Add(4 * fifthOfValidityDuration)
+	if r.currentTime.After(promoteSecondaryCATime) {
+		return CARotatePromoteSecondary, nil
+	}
+
+	// Delete expired secondary
+	if secondary != nil && r.currentTime.After(secondary.NotAfter) {
+		return CARotateDeleteSecondary, nil
+	}
+
+	return CARotateNoAction, nil
+}
+
+func (r *createCentralTLSExtensionRun) handleCARotation(fileMap types.SecretDataMap) (bool, error) {
+	reissueCentralCert := false
+	switch r.caRotationAction {
+	case CARotateAddSecondary:
+		ca, err := certgen.GenerateCA()
+		if err != nil {
+			return false, errors.Wrap(err, "creating secondary CA failed")
+		}
+		certgen.AddSecondaryCAToFileMap(fileMap, ca)
+
+	case CARotateDeleteSecondary:
+		certgen.RemoveSecondaryCA(fileMap)
+
+	case CARotatePromoteSecondary:
+		certgen.PromoteSecondaryCA(fileMap)
+		newPrimaryCa, err := certgen.LoadCAFromFileMap(fileMap)
+		if err != nil {
+			return false, errors.Wrap(err, "loading new primary CA failed")
+		}
+		r.ca = newPrimaryCa
+		reissueCentralCert = true // need to reissue because the primary CA has changed
+
+	case CARotateNoAction:
+		// no rotation was done, but if this function was called it means that validateAndConsumeCentralTLSData
+		// failed, so let's reissue the Central service certificate
```

Comment:
This assumption seems like something fragile. It's also a little confusing because IIRC `validateAndConsumeCentralTLSData` is called in two places. Would it be better to have this method only rely on the information it has, and have the caller reissue regardless of this method's  return value, if it deems it necessary? 🤔 

In fact, looking at `generateCentralTLSData`, it seems like up until now we reissued central cert unconditionally by calling `IssueCentralCert` there. Maybe we should keep doing it? The CA-related changes will be very rare so perhaps the additional complexity of computing whether a central cert update is needed (and the risk of missing some corner case where we skip it but in fact it's beneficial) is not worth it?

Reply:
I was also not sure sure about this one, and I considered your suggestion, before implementing it this way. What bothered me is that we'd recreate the Central TLS certificate for unrelated events, but then again there's nothing harmful in recreating the TLS certs, so might as well just do it and simplify the logic.

Reply:
I changed it to your suggestion.